nova_vm/src/ecmascript/execution/agent.rs

```diff
@@ -124,6 +123,130 @@ pub trait HostHooks: std::fmt::Debug {
     }
 }
 
+/// Owned ECMAScript Agent that can be used to run code but also to run garbage
+/// collection on the Agent heap.
+pub struct GcAgent {
+    agent: Agent,
+    realm_roots: Vec<Option<RealmIdentifier>>,
+}
+
+/// ECMAScript Realm root
+///
+/// As long as this is not passed back into GcAgent, the Realm it represents
+/// won't be removed by the garbage collector.
+#[must_use]
+#[repr(transparent)]
+pub struct RealmRoot {
```

Comment:
So the heap _does_ have the `globals` vector which is intended for rooting Values. I haven't yet really decided on how this all should work and only have vague ideas but I'm kind of thinking that the Jobs we give out from the Agent would either only contain rooted values, or would itself be a handle to an actual Job on the heap and you could then choose to root that.

I think it's maybe not worth it to have separate heap vectors (or these could be `HashMap<u32, T>`) for each rootable type: A root vector of Values, and a root vector of Jobs or promise reactions or something around that.

------

Yeah. I wasn't _too_ worried about this since the realm will eventually get collected together with the Agent itself. It IMO should be kind of okay to do:
```ts
let agent = GcAgent::new();
let realm = agent.create_realm();
drop(agent);
drop(realm);
```
ie. somehow accidentally mess up the drop order. If this panicked it would be in slightly poor taste.

Eventually I'd want to bind the RealmRoot lifetime to the GcAgent lifetime somehow, though I expect that isn't actually possible in an easy way. That lifetime binding would make the type a little less "annoying".

The ultimate thing would be a [linear type](https://faultlore.com/blah/linear-rust/), I think. This would basically be bit of an upscaled `#[must_use]` where the use must be (eventually) passing the value back to us. That would be nice.

------

For now, I think a panicking Drop impl is fine. It's a bit rough, but it definitely does the trick.

Reply:
Turns out a Drop impl is probably too rough after all. eg. You can no longer use the ? operator to deal with a result from `run_in_realm` nor call `exit(1)` within it: The Drop impl of that realm root will then panic with a wholly unrelated message. This also makes the test runner get quite confused.